src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -448,83 +504,64 @@ protected internal void OnSolutionReloaded(SolutionInfo reloadedSolutionInfo)
         /// </summary>
         protected internal void OnSolutionRemoved()
         {
+            // This currently doesn't use the SetCurrentSolution(transform) pattern as it changes mutable state (open
+            // docs), and as such needs to atomically change both that and the solution-snapshot.
+
             using (_serializationLock.DisposableWait())
             {
-                var oldSolution = this.CurrentSolution;
-
-                this.ClearSolutionData();
+                var (oldSolution, _) = this.ClearSolutionData_NoLock();
```

Comment:
Could you do the "clear" with your onAfterUpdate action? That's what I figured you needed it for.